modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -689,33 +696,48 @@ private CompletableFuture<Boolean> beforeZoneReplicaStarted(LocalPartitionReplic
                 .thenCompose(v -> {
                     ZonePartitionId zonePartitionId = parameters.zonePartitionId();
 
-                    Set<TableImpl> zoneTables = zoneTablesRawSet(zonePartitionId.zoneId());
-
-                    int partitionIndex = zonePartitionId.partitionId();
-                    PartitionSet singlePartitionIdSet = PartitionSet.of(partitionIndex);
-
-                    CompletableFuture<?>[] futures = zoneTables.stream()
-                            .map(tbl -> inBusyLockAsync(busyLock, () -> {
-                                return getOrCreatePartitionStorages(tbl, singlePartitionIdSet)
-                                        .thenRun(() -> {
-                                            localPartsByTableId.compute(
-                                                    tbl.tableId(),
-                                                    (tableId, oldPartitionSet) -> extendPartitionSet(oldPartitionSet, partitionIndex)
-                                            );
-                                        })
-                                        .thenRunAsync(() -> inBusyLock(busyLock, () -> {
-                                            lowWatermark.getLowWatermarkSafe(lwm ->
-                                                    registerIndexesToTable(tbl, catalogService, singlePartitionIdSet, tbl.schemaView(), lwm)
-                                            );
-
-                                            preparePartitionResourcesAndLoadToZoneReplica(tbl, zonePartitionId, false);
-                                        }), ioExecutor)
-                                        // If the table is already closed, it's not a problem (probably the node is stopping).
-                                        .exceptionally(ignoreTableClosedException());
-                            }))
-                            .toArray(CompletableFuture[]::new);
+                    StampedLock zoneLock = tablesPerZoneLocks.computeIfAbsent(zonePartitionId.zoneId(), id -> new StampedLock());
```

Comment:
Could you please switch to async RW locks? Long-term locks that block threads cause problems

Reply:
I'm ok with postponing this to next PR